### DISCUSSION TITLE
How to load modules conditionally depending of typeof window

### DISCUSSION THREAD

**Discussion by Smruti-AT** [webpack/webpack#19006] [Q&A]
I have created an SDK and which has some browser specific APIs and some Node specifc APIs. I want to configure the webpack bundle in such a way that those APIs are loaded only at runtime depending on the environment (web or node). I've tried dynamic importing and webpack conditional loaders but they didn't work and all the APIs are included in the bundle at compile time itself. Is there any way to achieve this behaviour?

PS: I'm trying the use the same file for both the environments

**Comment by alexander-akait** [comment]
This is not so easy to achieve and depends directly on how the code is written. You will need in any case to check the environment and load the necessary modules depending on the environment, also we have - https://github.com/webpack/webpack/issues/6525 (in the next release this will be improved), and be supported only for ES modules

**Reply by Smruti-AT** [reply]
So with webpack conditional loader, I had this if block
`/// #if type of window === 'undefined
// dynamically import browser APIs
// #endif
`
this doesn't work. Is there any way to check environments then?

**Reply by alexander-akait** [reply]
Using js code - `if (typeof window !== 'undefined') {   } else {   }`

**Reply by Smruti-AT** [reply]
I've tried this as well. In this case, always the else case is executed indicating that window is undefined. I think during bundling, 'window' is undefined

**Reply by alexander-akait** [reply]
We don't change env and don't touch `window`, check your code

**Reply by Smruti-AT** [reply]
this is my source file
![image](https://github.com/user-attachments/assets/eec1da50-641d-46d9-affb-842b0cf43575)
the error I'm getting when importing the code in a NodeJS file
![image](https://github.com/user-attachments/assets/ca419ffc-3901-4317-9434-058832a07fa9)
@mui shouldn't have loaded in the first place as it is used inside a browser specific component which itself is imported inside typeof window !== 'undefined' block, right?

this is the bundled code 
![image](https://github.com/user-attachments/assets/1a752609-fa33-457d-ad46-ec4737c063c5)

**Reply by alexander-akait** [reply]
> the error I'm getting when importing the code in a NodeJS file

This does not apply to webpack, use `mjs` extension or set `type: "module"`

**Reply by Smruti-AT** [reply]
the package.json of the SDK has `type: "module"` and in webpack.config.js the `output.library.type: "module"`. The package.json of the nodejs file is also `type: "module"`

**Reply by alexander-akait** [reply]
Can you create a reproducible repository using github to demonstrate the issue?

**Reply by Smruti-AT** [reply]
Here is the github repo: https://github.com/Smruti-AT/webpack_demo
though I got stuck with another error.
I had to move the sdk folder inside the src folder of the react app else it was giving an importing error

**Reply by alexander-akait** [reply]
I don't think your idea can be implemented using `Create React App`, because it is designed for use in a browser environment only

**Reply by Smruti-AT** [reply]
Initially, the SDK was not a part of the Create React App. It was an independent JS module. There I import different resource based on the environment (web or node). I had to move the sdk inside the src directory due to some errors.

**Reply by alexander-akait** [reply]
The problem is that the bundling of CRA is configured only for web target, we do not change the runtime code at all, but exclude global NODE variables - https://webpack.js.org/configuration/node/, therefore, all the logic is based on how the project itself is written.
